File: src/crawlers/lib/specific_crawlers/qualifying.py
import sys
from lib.tojson import exportJson
from lib.common import parse_url_content, get_driver_info, team_color

if __name__ == "__main__":
  year = sys.argv[1]
  location = sys.argv[2]
  id = sys.argv[3]
  sprint = False if len(sys.argv) < 5 else sys.argv[4] == "sprint"

  # Formula 1 webpage for qualifying results
  url = f"https://www.formula1.com/en/results.html/{year}/races/{id}/{location}/{"sprint-" if sprint else ""}qualifying.html"

  session_name, circuit_name, table = parse_url_content(url)

  results = []
  for row in table.find('tbody').find_all('tr'):
      columns = row.find_all('td')
      result = {
          "position": None if columns[1].text.strip() == "NC" else int(columns[1].text.strip()),
          "driver": get_driver_info(columns[3], columns[2]),
          "car": columns[4].text.strip(),
          "teamColor": team_color(columns[4].text.strip()),
          "times": {
            "q1": columns[5].text.strip() if columns[5].text.strip() else None,
            "q2": columns[6].text.strip() if columns[6].text.strip() else None,
            "q3": columns[7].text.strip() if columns[7].text.strip() else None,
          },
          "laps": int(columns[8].text.strip())
      }
      results.append(result)
  
  exportJson(
    data=results,
    id=id,
    location=location,
    circuit=circuit_name,
    session_name=session_name,
    data_type=f"sprint-qualifying" if sprint else "race-qualifying",
    exportname="sprint_qualifying" if sprint else "race_qualifying", 
    year=year
  )

File: src/crawlers/scripts/qualifying.ts
import {
  extractDriverInfo,
  findSessionInfo,
  getTeamColor,
  parseUrlContent,
} from "../lib/common";
import { writeData } from "../lib/export";
import { URL_BASE } from "../lib/constants";

interface QualifyingResult {
  position: number;
  driver: Driver;
  car: string;
  teamColor: string;
  times: { q1: string; q2: string; q3: string };
  laps: number;
}

interface Params {
  year: number | string;
  id: string;
  location: string;
  isSprint?: boolean;
}

export async function fetchQualifyingResults({
  year,
  id,
  location,
  isSprint,
}: Params) {
  try {
    const url = `${URL_BASE}/${year}/races/${id}/${location}/${
      isSprint ? "sprint-qualifying" : "qualifying"
    }.html`;

    const { $, sessionName, circuit, tableRows } = await parseUrlContent(url);

    const sprintInSessName = sessionName.toLocaleLowerCase().includes("sprint");
    if (!sprintInSessName && isSprint) return null;
    if (sprintInSessName && !isSprint) return null;

    const results: QualifyingResult[] = [];

    tableRows.each((_, row) => {
      const position = parseInt(
        $(row).find("td:nth-child(2)").text().trim(),
        10
      );
      const number = parseInt($(row).find("td:nth-child(3)").text().trim(), 10);
      const rawDriver = $(row)
        .find("td:nth-child(4)")
        .text()
        .trim()
        .replace(/\s+/g, " ");
      const driver = { ...extractDriverInfo(rawDriver), number };
      const car = $(row).find("td:nth-child(5)").text().trim();
      const teamColor = getTeamColor(car) ?? "";
      const times = {
        q1: $(row).find("td:nth-child(6)").text().trim() ?? null,
        q2: $(row).find("td:nth-child(7)").text().trim() ?? null,
        q3: $(row).find("td:nth-child(8)").text().trim() ?? null,
      };
      const laps = parseInt($(row).find("td:nth-child(9)").text().trim(), 10);

      results.push({ position, driver, car, teamColor, times, laps });
    });

    writeData(results, {
      id,
      dataType: isSprint ? "sprint-qualifying" : "race-qualifying",
      year,
      url,
      sessionName,
      circuit,
      location,
      fileName: `${isSprint ? "sprint" : "race"}_qualifying`,
    });
    return results;
  } catch (error) {
    console.error("Error fetching the data:", error);
  }
}
